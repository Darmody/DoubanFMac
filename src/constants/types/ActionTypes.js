// auth
export const LOGIN_REQUEST: ActionType = 'LOGIN/REQUEST'
export const LOGIN_SUCCESS: ActionType = 'LOGIN/SUCCESS'
export const LOGIN_FAILURE: ActionType = 'LOGIN/FAILURE'

// users
export const USER_CURRENT_REQUEST: ActionType = 'USER/CURRENT/REQUEST'

// entities
export const ENTITIES_SAVE: ActionType = 'ENTITIES/SAVE'
export const ENTITIES_UPDATE: ActionType = 'ENTITIES/UPDATE'

// songs
export const SONG_CURRENT: ActionType = 'SONGS/CURRENT'
export const SONG_LISTEN_REQUEST: ActionType = 'SONGS/LISTEN/REQUEST'
export const SONG_NEXT_REQUEST: ActionType = 'SONGS/NEXT/REQUEST'
export const SONG_MARK_REQUEST: ActionType = 'SONGS/MARK/REQUEST'
export const SONG_MARK_SUCCESS: ActionType = 'SONGS/MARK/SUCCESS'
export const SONG_LIKE_REQUEST: ActionType = 'SONGS/LIKE/REQUEST'
export const SONG_LIKE_SUCCESS: ActionType = 'SONGS/LIKE/SUCCESS'
export const SONG_DISLIKE_REQUEST: ActionType = 'SONGS/DISLIKE/REQUEST'
export const SONG_DISLIKE_SUCCESS: ActionType = 'SONGS/DISLIKE/SUCCESS'
