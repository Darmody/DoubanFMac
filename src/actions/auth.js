// @flow
import * as types from 'constants/types/ActionTypes'
import type { Action } from 'constants/types/Redux'

export const login = (username: string, password: string): Action => ({
  type: types.LOGIN_REQUEST,
  payload: {
    username,
    password,
  }
})

export const logout = (): Action => ({
  type: types.LOGOUT,
  payload: undefined,
})

export const logined = (response: Object): Action => ({
  type: types.LOGIN_SUCCESS,
  payload: response,
})
