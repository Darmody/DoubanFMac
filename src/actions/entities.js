// @flow
import * as types from 'constants/types/ActionTypes'
import type { Action } from 'constants/types/Redux'

export const save = (data: Object): Action => ({
  type: types.ENTITIES_SAVE,
  payload: data,
})

export const update = (data: Object): Action => ({
  type: types.ENTITIES_UPDATE,
  payload: data,
})
