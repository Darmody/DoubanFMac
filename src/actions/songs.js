// @flow
import * as types from 'constants/types/ActionTypes'
import type { Action } from 'constants/types/Redux'

export const listen = (channel: number): Action => ({
  type: types.SONG_LISTEN_REQUEST,
  payload: channel,
})

export const next = (channel: number, sid?: number): Action => ({
  type: types.SONG_NEXT_REQUEST,
  payload: {
    channel,
    sid,
  }
})

export const mark = (channel: number, sid: number): Action => ({
  type: types.SONG_MARK_REQUEST,
  payload: {
    sid,
    channel,
  }
})

export const current = (song: {}): Action => ({
  type: types.SONG_CURRENT,
  payload: song,
})
