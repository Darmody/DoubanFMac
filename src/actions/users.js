// @flow
import * as types from 'constants/types/ActionTypes'
import type { Action } from 'constants/types/Redux'

export const current = (): Action => ({
  type: types.USER_CURRENT_REQUEST,
  payload: undefined,
})
