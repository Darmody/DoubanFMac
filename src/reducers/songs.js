// @flow
import Immutable from 'seamless-immutable'
import * as types from 'constants/types/ActionTypes'
import type { Reducer } from 'constants/types/Redux'

const initialState = Immutable({
  id: undefined,
  playedIds: [],
})

const reducer: Reducer = (state = initialState, { type, payload = {} }) => {
  switch (type) {
    case types.SONG_CURRENT: return state.merge({
      id: payload.sid,
    })
    case types.SONG_PLAYED_SUCCESS: return state.merge({
      playedIds: payload,
    })
    default: return state
  }
}

export default reducer
