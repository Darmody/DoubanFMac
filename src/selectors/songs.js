// @flow
import { SONG } from 'schemas'
import type { StateShape } from 'constants/types/Redux'
import { selectEntity } from 'selectors/entities'

export const selectCurrent = (state: StateShape) => selectEntity(
  state, SONG, state.songs.id,
)
