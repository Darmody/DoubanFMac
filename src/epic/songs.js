// @flow
import { Observable as Rx$ } from 'rxjs'
import * as types from 'constants/types/ActionTypes'
import type { Epic } from 'constants/types/Redux'
import * as API from 'clients/doubanRxClient'
import { current } from 'actions/songs'
import { update as updateEntities } from 'actions/entities'
import { normalizeResponse, getToken, fullfilled } from 'utils/operators'
import { SONG } from 'schemas'

const fetchSong = (type, api) => (action$, store) => action$
  .ofType(type)
  .pluck('payload')
  .switchMap(api(getToken(store)))
  .pluck('response', 'song')
  .mergeMap(([song]) => Rx$.merge(
    Rx$.of(normalizeResponse(song, SONG)),
    Rx$.of(current(song)),
  ))

const nextEpic: Epic = fetchSong(types.SONG_NEXT_REQUEST, API.nextSong)
const listenEpic: Epic = fetchSong(types.SONG_LISTEN_REQUEST, API.listenSong)
const banEpic: Epic = fetchSong(types.SONG_BAN_REQUEST, API.banSong)

const markEpic: Epic = (action$, store) => action$
  .ofType(types.SONG_MARK_REQUEST)
  .pluck('payload')
  .mergeMap(API.markSong(getToken(store)))
  .map(fullfilled(types.SONG_MARK_SUCCESS))

const likeEpic: Epic = (action$, store) => {
  const likeFullfilled = sid => () => Rx$.of(updateEntities([{
    field: ['songs', sid, 'like'],
    value: 1,
  }]))

  return action$
    .ofType(types.SONG_LIKE_REQUEST)
    .pluck('payload')
    .mergeMap(payload => API
      .likeSong(getToken(store))(payload)
      .mergeMap(likeFullfilled(payload.sid))
    )
}

const dislikeEpic: Epic = (action$, store) => {
  const dislikeFullfilled = sid => () => Rx$.of(updateEntities([{
    field: ['songs', sid, 'like'],
    value: 0,
  }]))

  return action$
    .ofType(types.SONG_DISLIKE_REQUEST)
    .pluck('payload')
    .mergeMap(payload => API
      .dislikeSong(getToken(store))(payload)
      .mergeMap(dislikeFullfilled(payload.sid))
    )
}

export default [
  banEpic,
  dislikeEpic,
  likeEpic,
  listenEpic,
  markEpic,
  nextEpic,
]
