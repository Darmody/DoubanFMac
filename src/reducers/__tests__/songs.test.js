import * as types from 'constants/types/ActionTypes'
import * as actions from 'actions/songs'
import reducer from 'reducers/songs'

describe('reducer::songs', () => {
  it('initialize state', () => {
    expect(reducer(undefined, {})).toMatchSnapshot()
  })

  it('切换当前歌曲', () => {
    const state = reducer(undefined, actions.current({ sid: 1 }))
    expect(state).toMatchSnapshot()
    expect(state.id).toBe(1)
  })

  it('记录已播放的歌曲 ids', () => {
    const playedIds = [1, 2]
    const state = reducer(
      undefined,
      { type: types.SONG_PLAYED_SUCCESS, payload: playedIds },
    )
    expect(state).toMatchSnapshot()
    expect(state.playedIds).toEqual(playedIds)
  })
})
