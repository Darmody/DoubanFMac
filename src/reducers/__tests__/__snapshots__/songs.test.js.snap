// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer::songs initialize state 1`] = `
Object {
  "id": undefined,
  "playedIds": Array [],
}
`;

exports[`reducer::songs 切换当前歌曲 1`] = `
Object {
  "id": 1,
  "playedIds": Array [],
}
`;

exports[`reducer::songs 记录已播放的歌曲 ids 1`] = `
Object {
  "id": undefined,
  "playedIds": Array [
    1,
    2,
  ],
}
`;
